src/emulation/promises.ts

```diff
@@ -656,13 +656,34 @@ export async function mkdir(path: fs.PathLike, options: fs.MakeDirectoryOptions
 export async function mkdir(path: fs.PathLike, options?: fs.Mode | (fs.MakeDirectoryOptions & { recursive?: false | undefined }) | null): Promise<void>;
 export async function mkdir(path: fs.PathLike, options?: fs.Mode | fs.MakeDirectoryOptions | null): Promise<string | undefined>;
 export async function mkdir(path: fs.PathLike, options?: fs.Mode | fs.MakeDirectoryOptions | null): Promise<string | undefined | void> {
+	options = typeof options === 'object' ? options : { mode: options };
+	const mode = normalizeMode(options?.mode, 0o777);
+
 	path = normalizePath(path);
 	path = (await exists(path)) ? await realpath(path) : path;
 	const { fs, path: resolved } = resolveMount(path);
-	try {
-		await fs.mkdir(resolved, normalizeMode(typeof options == 'object' ? options?.mode : options, 0o777), cred);
-	} catch (e) {
-		throw fixError(e as Error, { [resolved]: path });
+	const errorPaths: Record<string, string> = { [resolved]: path };
+
+	const mkdirSingle = async (dir: string) => {
+		try {
+			await fs.mkdir(dir, mode, cred);
+		} catch (e) {
+			throw fixError(e as Error, errorPaths);
+		}
+	};
+
+	if (options?.recursive) {
+		const dirs: string[] = [];
+		for (let dir = resolved, origDir = path; !(await fs.exists(dir, cred)); dir = dirname(dir), origDir = dirname(origDir)) {
+			dirs.unshift(dir);
+			errorPaths[dir] = origDir;
+		}
+		for (const dir of dirs) {
+			await mkdirSingle(dir);
+		}
+		return dirs[0];
+	} else {
+		return mkdirSingle(resolved);
```

Comment:
This will not handle multiple mounts correctly. In addition, I find it to be very difficult to read. I think it would be much easier to recursively call `mkdir`. For example, you could prefix the call to `FileSystem.mkdir` with:

```ts
const parent = dirname(path);
if(!(await exists(parent))) {
     await mkdir(parent, options);
}
```

This correctly handles multiple mounts, and is a lot simpler than the other implementation.

Reply:
I don't understand the concern regarding multiple mounts.
In Unix filesystem, the mountpoint directory must always exists, so that `mkdir -p` command can only operate within one mountpoint.
Is it working any differently here?

The solution you suggested would not have the correct return value, which shall be the first directory created.

Reply:
That makes sense- not sure what I was thinking. 

As for the return value- the code I showed is just an example.

Reply:
There are several more complications than the return value.
The other issue is, mkdir an existing directory is considered successful in recursive mode, not throws EEXIST in non-recursive mode. 
With these conditions tacked on, the alternative solution would not be cleaner.

Reply:
I agree, thanks again. I made some minor syntax changes to improve performance and readability.